File: game/pieces.go
package game

type Color byte
type Kind byte
type Piece byte

const (
	White Color = 1 << 7
	Black Color = 0 << 7

	colorMask byte = 1 << 7

	Flat     Kind = 1
	Standing Kind = 2
	Capstone Kind = 3

	typeMask byte = 1<<2 - 1
)

func MakePiece(color Color, kind Kind) Piece {
	return Piece(byte(color) | byte(kind))
}

func (p Piece) Color() Color {
	return Color(byte(p) & colorMask)
}

func (p Piece) Kind() Kind {
	return Kind(byte(p) & typeMask)
}

func (p Piece) IsRoad() bool {
	return p.Kind() == Flat || p.Kind() == Capstone
}

func (p Piece) String() string {
	c := ""
	if p.Color() == White {
		c = "W"
	} else {
		c = "B"
	}
	switch p.Kind() {
	case Capstone:
		c += "C"
	case Standing:
		c += "S"
	}
	return c
}

func (c Color) String() string {
	if c == White {
		return "white"
	}
	return "black"
}

func (c Color) Flip() Color {
	if c == White {
		return Black
	}
	return White
}

File: python/tak/pieces.py
import enum

from attrs import define


class Color(enum.Enum):
    WHITE = 0
    BLACK = 1

    def flip(self):
        return Color(1 - self.value)


class Kind(enum.Enum):
    FLAT = 0
    STANDING = 1
    CAPSTONE = 2

    def is_road(self):
        return self == Kind.FLAT or self == Kind.CAPSTONE


_piece_cache = [[None for k in Kind] for c in Color]


@define(frozen=True, slots=True)
class Piece(object):
    color: Color
    kind: Kind

    def is_road(self):
        return self.kind.is_road()

    @classmethod
    def _init_cache(cls):
        for c in Color:
            for k in Kind:
                _piece_cache[c.value][k.value] = cls(c, k)

    @classmethod
    def cached(self, color, kind):
        return _piece_cache[color.value][kind.value]


Piece._init_cache()

__all__ = ["Color", "Kind", "Piece"]
